doc/instead.6
.TH INSTEAD 6 "Version 3.3.2" Instead GAMES

.SH NAME

INSTEAD - Simple Text Adventure Engine, The Interpreter

.SH DESCRIPTION

The interpreter of STEAD (Simple Text Adventures) allows one to play games that combine visual novels, text adventures and classic quests of 1990-s.
The STEAD games features are:
.TP
 * very simple source code of the stories. (based on LUA);
.TP
 * graphical game interface;
.TP
 * supports music and images;
.TP
 * theme support \- a game may change the interface look;
.TP
 * portability (originally written for Linux, depends on SDL and lua).

.SH SYNOPSIS

.B sdl-instead
.I [options] [game.zip or game.idf or path to game]

.SH OPTIONS

.TP
.B -debug
Debug mode (for game developers).
.TP
.B -nosound
Run the game without sound.
.TP
.B -vsync
Vsync output. (SDL2 only)
.TP
.B -gamespath <path>
Add path with games.
.TP
.B -themespath <path>
Add path with themes.
.TP
.B -nostdgames
Do not use default games.
.TP
.B -nostdthemes
Do not use default themes.
.TP
.B -game <basename>
Select game in gamespath.
.TP
.B -theme <basename>
Select theme in themespath.
.TP
.B -owntheme
Force game to use own theme.
.TP
.B -notheme
Do not use own game theme.
.TP
.B -fullscreen
Run the game in fullscreen mode.
.TP
.B -nocursor
Do not use game mouse cursors (use system one)
.TP
.B -hires / -nohires
Set/unset the high resolution when run in fullscreen mode.
.TP
.B -window
Run the game in windowed mode.
.TP
.B -resizable
Resizable window (SDL2 only)
.TP
.B -scale <scale>
Scale window (SDL2 only)
.TP
.B -noautosave
Disable autosave load.
.TP
.B -encode <lua file> [encoded file]
Encode lua file. Encoded file can be executed with doencfile("encoded file").
.TP
.B -idf <path to resources directory or game>
Build idf file [INSTEAD data file] from directory.
.TP
.B -mode [WxH]
Use WxH resolution.
.TP
.B -modes <WxH>[,<WxH>[,<WxH>]] ...
Define available resolutions.
.TP
.B -software
Force software rendering
.TP
.B -renderer <opengl|direct3d|software|...>
Select default renderer (SDL2 only)
.TP
.B -fontscale <percents>
Font scaling.
.TP
.B -nopause
Do not pause the game on window minimize.
.TP
.B -lang <lang code>
Select language.
.TP
.B -hinting 0|1|2|3
Set the font hinting mode (helpful with infinality).
.TP
.B -install [game in zip]
Install game from zip archive
.TP
.B -quit
Quit :)
.TP
.B -appdata [fullpath]
Store saves and settings in appdata path. Path must exist!
.TP
.B -chunksize [size in bytes]
Size for audio buffer. Try this if sound lags.
.TP
.B -version
Show version and exit.
.TP
.B -lua <script.lua> [arguments]
Exec lua script (for game developers).
.TP
.B -luac <script.lua>
Check lua script (for game developers).
.TP
.B -standalone
Run game in standalone mode (no sandbox, no games selection ...).
.TP
.B -noconfig
Do not load saved config (run with defaults).
.TP
.B -profile <profile>
Using configuration profile. Profile file is the ini-styled file with parameter = value pairs. The parameters
are arguments without -.

.PP
Configuration file name is 'insteadrc' and it's located in ~/.instead/.

Games can be downloaded from http://instead-games.sf.net. Unzip them
in ~/.instead/games.

.SH CONTROLS

Please, see tutorial.

.SH LICENCE

This software is covered by the terms of the MIT license.

.SH WARRANTY
None.

.SH COPYRIGHT AND CREDITS

.TP
.B Peter Kosyh
INSTEAD author.
.TP
.B Ilya Ryndin
Initial port to Windows.
.TP
.B Boris Timofeev, Anton Kolosov, Aleksey Mahno
Port to Android.
.TP
.B .dm
Port to Mac OS X.
.TP
.B Andrey Afletdinov
Port to Windows CE.
.TP
.B Vadim Balashoff
Advice.
.TP
.B Vladimir Zhirov
Advice.
.TP
.B Alexander Yakovlev
manual.tex support.
.TP
.B Serj Kalichev
Advice and default theme correction.
.TP
.B Vladimir Podobaev, Oleg Gvozdev
Testers.
.TP
.B Alexander Soborov
Wiki support.
.TP
.B Andrew Lobanov
Games repository, slackbuild.
.TP
Many others for comments, bug reports, games and ideas.

.SH SEE ALSO
.BR https://instead-hub.github.io


CHANGES
1.4.5:

  * bug fix in rnd() w/o parameter;
  * callpush/callpop/cctx/strip moved to stead;
  * bug fix in for_each;
  * bug fix in txtnb and \;
  * win.align added to theme;
  * theme_name added;
  * now fading first value is new scene flag;
  * debian files cleanup;

 -- Peter Kosyh <p.kosyh@gmail.com>  Thu, 14 Jul 2011 16:18:00 +0400

1.4.4:

  * bug fix in direct mode (memory leak);

 -- Peter Kosyh <p.kosyh@gmail.com>  Thu, 21 Jun 2011 08:57:00 +0400

1.4.3:

  * bug fix in nopara logic;
  * ukranian translation and tutorial;
  * multilang tutorial3;
  * sprites small fix (predefined handles);
  * cursor changing from direct mode support;

 -- Peter Kosyh <p.kosyh@gmail.com>  Thu, 19 Jun 2011 10:24:00 +0400

1.4.2:

  * bug fix in idf gets;
  * bug fix in kbd hooking;

 -- Peter Kosyh <p.kosyh@gmail.com>  Thu, 19 May 2011 20:14:00 +0400

1.4.1:

  * bug fix in use;

 -- Peter Kosyh <p.kosyh@gmail.com>  Thu, 12 May 2011 23:53:00 +0400

1.4.0:

  * bug fix in imgl/imgr;
  * bug fix in xact;
  * bug fix in jump to pos logic;
  * bug fix lags in sound;
  * bug fix with wince/windows with themes load;
  * experemental sprites technology;
  * global dir is now readdir;
  * multichannal sound system (add_sound, stop_sound);
  * -appdata parameter;
  * -chunksize parameter;
  * show dir to be deleted while remove game;
  * theme reset features;
  * mouse_pos added;
  * start() init function added;
  * idf files support (instead data format);
  * menu font is now scalable too;
  * toggle_menu added;
  * get_ticks added;
  * bit_xxx bitwise operations added;
  * visits() added;

 -- Peter Kosyh <p.kosyh@gmail.com>  Mon, 2 May 2011 08:35:00 +0400  

1.3.4:

  * bug fix in text renderer with italic text;
  * bug fix with imgl/imgr in inv;
  * bug fix in video init;
  * bug fix in theme.get 'inv.mode'
  * font faces { } syntax;
  * new default font;
  * set light hinting for font;
  * added stead.dir iterator;
  * modules doc fix;
  * languages dir renamed to lang;
  * updated manual.pdf;

 -- Peter Kosyh <p.kosyh@gmail.com>  Fri, 11 Mar 2011 12:10:00 +0300  

1.3.3:

  * bug fix (kbd input);
  * modules docs;

 -- Peter Kosyh <p.kosyh@gmail.com>  Thu, 27 Feb 2011 20:27:00 +0300  

1.3.2:

  * bug fix (dates in save slots);
  * bug fix (dbg fixes);
  * bug fix (f8 and f9 error msg);
  * small bug fixes;
  * s60 build;
  * SDL 1.3 ready;
  * it lang and tutorial;

 -- Peter Kosyh <p.kosyh@gmail.com>  Thu, 24 Feb 2011 13:12:00 +0300  

1.3.1:

  * bug fix (imgl and justify);
  * bug fix (languages in WinCE version);
  * bug fix (prefs do not create save dir);
  * bug fix (alt+f4 != f4);
  * bug fix (<w:> tag parsing)
  * visual.lua moved to doc/;

 -- Peter Kosyh <p.kosyh@gmail.com>  Mon, 29 Nov 2010 15:44:00 +0300  

1.3.0:

  * bug fix (resample sounds while HZ change);
  * bug fix (empty bg);
  * bug fix (cursor center scaling);
  * changing themes from game;
  * strike ougth text;
  * *.fnt.height theme parameter;
  * scroller positions in theme;
  * escaping ^ and delim;
  * box: and blank:;
  * pad: ;
  * speed up;
  * picture flow;
  * prefs:purge now removes vars;
  * dialog:empty added;
  * txttab;
  * tutorial update;

 -- Peter Kosyh <p.kosyh@gmail.com>  Tue, 12 Nov 2010 22:42:00 +0300  

1.2.3:

  * android build;
  * bug in xact;

 -- Peter Kosyh <p.kosyh@gmail.com>  Sat, 2 Oct 2010 16:53:00 +0300  

1.2.2:

  * xact now can be used from everywhere;
  * much speed improvments;
  * get_gamepath, get_steadpath;
  * wince port;
  * internal cleanups;

 -- Peter Kosyh <p.kosyh@gmail.com>  Fri, 25 Sep 2010 13:59:00 +0300  

1.2.1:

  * clearlooks bg;
  * fading fix;
  * fix in disable autosave;
  * disp fix;
  * savevars optimization;

 -- Peter Kosyh <p.kosyh@gmail.com>  Sat, 28 Aug 2010 19:47:00 +0300  

1.2.0:

  * unpack/install feature;
  * browse feature (win32 and gtk);
  * clever game/themes sorting;
  * time of save slots;
  * remove games;
  * aligned width for themes/games dialog;
  * local appdata mode (portable app);
  * clearlooks theme;
  * inv align modes;
  * prefs object;
  * init() function;
  * snapshots (not documented?);
  * hook and inherit;
  * entered/left human friendly actions;
  * stead.cmd, stead.args added;
  * require added (modules: goto, xact, input, click, vars, dbg, snapshot, prefs, format, kbd, hotkeys);
  * instead_version added;
  * taketo/takef added;
  * disable/enable/disable_all/enable_all added;
  * exist() added (seen over disabled objects);
  * path() added;
  * visited() added;
  * live() added;
  * nameof() added;
  * goback() added;
  * goin()/goout() added;
  * disp attribute added;
  * improved debugger;
  * save vars in _G;
  * no duplicated games while looking gamespaths;
  * code function!!!;
  * args in call;
  * var and global;
  * self() is now self;
  * psen/punseen for dlg;
  * no return goto needed;
  * disable_all/enable_all for lists;
  * purge and list_purge;
  * undocumented gamefile;
  * undocumented LANG;
  * many bugfixes...
  * mac os x port

 -- Peter Kosyh <p.kosyh@gmail.com>  Wed, 04 Aug 2010 20:59:00 +0300  

1.1.6:

  * bugfix (no closed files)
  * bugfix in line breaking with gfx
  * pr() function added in stead

 -- Peter Kosyh <p.kosyh@gmail.com>  Wed, 19 May 2010 17:31:00 +0300  

1.1.5:

  * multiple anigif in one layout fix
  * click in picture event
  * game.action callback
  * bug in <g: > parsing
  * pause game while minimize (-nopause option added) 
  * 8 bit scaler fix

 -- Peter Kosyh <p.kosyh@gmail.com>  Wed, 10 Mar 2010 15:20:51 +0300  

1.1.4:

  * keyboard navigation fix
  * cursor while menu and lost focus
  * lower cpu usage (no gif logic if no gifs are drawn)

 -- Peter Kosyh <p.kosyh@gmail.com>  Fri, 05 Mar 2010 18:28:40 +0300  

1.1.3:

  * unix_path in get_img
  * " in variable names (save)
  * \\, in ways and inv
  * 32 bit bmp fix

 -- Peter Kosyh <p.kosyh@gmail.com>  Wed, 16 Feb 2010 11:10:40 +0300  

1.1.2:

  * bug in saves


src/tiny/instead.lua
-- pure Lua instead (stead3 api)

-- metaparser example:
--
-- local instead = require "instead":new()
-- print(instead:start())
-- for l in io.lines() do
--   print(instead:parser(l))
-- end

local env = {
  __super = false,
  __mods_loaded__ = {},
  package = { path = './stead/stead3/?.lua;'..package.path },
  debug = debug,
  type = type,
  setfenv = setfenv,
  getfenv = getfenv,
  rawset = rawset,
  rawget = rawget,
  rawequal = rawequal,
  setmetatable = setmetatable,
  getmetatable = getmetatable,
  table = table,
  math = math,
  string = string,
  pairs = pairs,
  ipairs = ipairs,
  io = io,
  os = os,
  tonumber = tonumber,
  tostring = tostring,
  print = print,
  collectgarbage = collectgarbage,
  error = error,
  sys = sys,
  PLATFORM = 'unix',
  DEBUG = false,
}

if not table.clone then
  function table.clone(src)
    local dst = {}
    if type(src) ~= 'table' then return src end
    for k, _ in pairs(src) do
      dst[table.clone(k)] = table.clone(src[k])
    end
    return dst
  end
end

if not string.split then
  function string.split(self, n, sep, rexp)
    if type(n) ~= 'number' then
      sep, rexp, n = n, sep, false
    end
    if not sep then
      rexp = true
      sep = "[ \t]+"
    end
    local ret = {}
    if self:len() == 0 then
      return ret
    end
    n = n or -1
    local idx, start = 1, 1
    local s, e = self:find(sep, start, not rexp)
    while s and n ~= 0 do
      ret[idx] = self:sub(start, s - 1)
      idx = idx + 1
      start = e + 1
      s, e = self:find(sep, start, not rexp)
      n = n - 1
    end
    ret[idx] = self:sub(start)
    return ret
  end
end

function env.instead_gamepath()
  return "./"
end

local loadf = loadfile

local function make_loadfile(n, env)
  if setfenv then
    setfenv(1, env)
  end
  return loadf(n, "t", env)
end

local function make_dofile(n, env)
  local r, e = make_loadfile(n, env)
  if setfenv then
    setfenv(1, env)
  end
  if not r then
    error(e..'\n'..debug.traceback())
  end
  return r()
end

local function make_require(n, env)
  local mods = env.__mods_loaded__
  if mods[n] then
    return mods[n]
  end
  local pathes = env.package.path:split(";")
  for _, p in ipairs(pathes) do
    local name = p:gsub("%?", n)
    local f = io.open(name, "r")
    if f then
      f:close()
      mods[n] = make_dofile(name, env) or true
      break
    end
  end
  if not mods[n] then
    error("Can't load module: "..tostring(n), 2)
  end
  return mods[n]
end


local function init(env, f)
  local _ENV = env
  if setfenv then
    setfenv(1, _ENV)
  end
  require("stead")
  require("tiny3")
  stead:init()
  std.mod_start(function()
    local mp = std.ref '@metaparser'
    if mp then
      VerbExtend ({
        "#MetaSave",
        "*:MetaSave",
      }, mp)
      VerbExtend ({
        "#MetaLoad",
        "*:MetaLoad",
      }, mp)
      rawset(mp, 'MetaSave', function(self, w)
        __super:save(w)
      end)
      rawset(mp, 'MetaLoad', function(self, w)
        __super.need_load = w
      end)
      rawset(instead, 'restart', function()
        __super.need_restart = true
      end)
    end
  end)
  dofile(f or "main3.lua")
end

local boot = {}

function boot:newenv()
  local new = table.clone(env) -- mt can't be used :(
  new.dofile = function(f) return make_dofile(f, new) end
  new.require = function(f) return make_require(f, new) end
  new.loadfile = function(f) return make_loadfile(f, new) end
  new._G = new
  init(new)
  return new
end

function boot:start()
  self.env.game:ini()
  return self.env.iface:cmd 'look'
end

function boot:restart()
  self.env = boot:newenv()
  self.env.__super = self
  return self:start()
end

function boot:load(file)
  self.env = boot:newenv()
  self.env.__super = self
  self.env.game:ini()
  return self.env.iface:cmd(string.format('load %s', file))
end

function boot:save(file)
  return self.env.iface:cmd(string.format('save %s', file))
end

function boot:cmd(cmd)
  return self.env.iface:cmd(cmd)
end

function boot:parser(cmd)
  return self:cmd(string.format("@metaparser %q", cmd))
end

function boot:new()
  local new = boot:newenv()
  local o = { env = new }
  new.__super = o
  self.__index = self
  setmetatable(o, self)
  return o
end

return boot
